### ISSUE TITLE
Is the d-vector extracted strictly according to the original d-vector paper?

### ISSUE THREAD

**Issue by qwfy** [mravanelli/SincNet#10]
**Table 2** of the paper compares the performance on the speaker verification task bewteen SincNet with d-vector and several other models. How's the d-vector extracted? Which of the following is correct?

1. The naive way. Namely a random chunk (200 ms, not necessarily comes from the same utterance) of the speaker's audio is pre-processed (`\x -> x / abs(max(y))`, where `x` is the current chunk and `y` is whole signal), and then fed to the network, the output of the last hidden layer (i.e. the output of `DNN1_net`) is obtained, do the same for other audio chunks (thus many audio chunks are consumed), one would get many vectors, denoted `d_1, d_2, ..., d_n`, each `d_i` is L2 normalized to get `d'_i` and the final d-vector of this speaker is `mean(d'_1, d'_2, ..., d'_n)`.

2. The "original" way. A speaker is represented by a sequence of utterance, `{O_i: i}`, each utterance is consisted of a sequence of frames, `{o_j: j}`, during enrollment/verification, each `o_j` is companied by its context, (the original paper uses 30 frames to the left and 10 frames to the right), and fed to the network, the output of the last hidden layer is extracted, denoted with `a_j`, `a_j` is then L2 normalized, the d-vector of utterance `O_i` is `sum_{j} {a_j}`; the d-vector of the speaker is `mean({d-vector of O_i: i})`

The d-vector paper: https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/41939.pdf

PS. in both cases, I assume that the network used for extracting d-vector is the same one that is produced by `speaker_id.py`.

**Comment by mravanelli** [comment]
Hi,
it's similar to 2 (our setup is also similar to that described in DEEP
NEURAL NETWORKS FOR SMALL FOOTPRINT TEXT-DEPENDENT SPEAKER VERIFICATION).
The difference is that we don't expand the context (since our input is
already a chunk of signal of 200 ms). Moreover, our i-vector is the average
of all activations, and not the sum (very minor difference).

**Comment by qwfy** [comment]
Thanks :+1:

**Comment by seandickert** [comment]
Hi,

Just for clarification, to extract a d-vector from a single signal or utterance (referencing the SincNet config file in the repo), we take N 200 ms chunks, with 10ms overlap, of the entire signal, where N is the number of chunks we get by walking through the entire signal.  For each chunk, we run it through the model, extracting the activation of the last layer in the dense network (should be of size 2048), resulting in N activations of length 2048 (N d-vectors for that utterance).  Finally, we apply an L2 norm across the N vectors, followed by taking the mean of them, resulting in 1 d-vector of length 2048?  Is this accurate?

I'm curious how this approach might be impacted by utterances which contain a large amount of non-speech noise between speech.  Will it effectively "water down" the speech samples by averaging it with the d-vectors of the non-speech noise?  Specifically, I'm talking about noise which is difficult to remove from an utterance using off the shelf VAD.  If using cosine scoring to compare d-vectors, I understand that it requires some method for aggregating the vectors so that we can compare utterances of different length without having to cross score many d-vectors or otherwise average scores somehow.  But, given that something like a PLDA model can use a distribution of a speaker's d-vectors to train and predict, would it be better to not normalize and aggregate per utterance and rather feed all of the N d-vectors to the PLDA model?

**Comment by mravanelli** [comment]
Hi,
yes, your description of d-vector extraction is correct.
I didn't consider in my study the presence of noise. It can clearly make
learning speaker identities much more difficult. One can also consider an
attention model (as far as I remember someone has used it for speaker
id/verification) that attributes higher weight to the most important
frames, and lower to the frames which are considered noisy.

Mirco
